spec/rspec/matchers/dsl_spec.rb

```diff
@@ -563,6 +563,28 @@ def foo
         expect(self.matcher.matches?('other value')).to be_falsey
       end
 
+      context "with a negated match block" do
+        let(:matcher) do
+          new_matcher(:name, "value") do |expected|
+            match do |actual|
+              expect(actual).to eq expected
+            end
+
+            match_when_negated do |actual|
+              expect(actual).not_to eq expected
+            end
+          end
+        end
+
+        it "returns true if the wrapped expectation passes" do
+          expect(self.matcher.does_not_match?('other value')).to be_truthy
+        end
```

Comment:
I think I figured out why this spec is failing -- I think the negation on [this line](https://github.com/rspec/rspec-expectations/blob/1c63b42d1ee1a07c14f9543e8274cc892f829031/lib/rspec/expectations/handler.rb#L72) is wrong.  Note, however, that removing that negation makes another spec fail, but I think that other spec is incorrect.

All that said: removing the negation on that line worries me a little as it's hard to keep track of all the negations going on in my head and I'm not 100% sure that's correct.  What would really help solidify it is to make sure we have specs covering all 8 of these cases:

* A matcher that uses a positive expectation in `match` checked against a matching value
* A matcher that uses a positive expectation in `match` checked against an unmatching value
* A matcher that uses a negative expectation in `match` checked against a matching value
* A matcher that uses a negative expectation in `match` checked against an unmatching value
* A matcher that uses a positive expectation in `match_when_negated` checked against a matching value
* A matcher that uses a positive expectation in `mach_when_negated` checked against an unmatching value
* A matcher that uses a negative expectation in `match_when_negated` checked against a matching value
* A matcher that uses a negative expectation in `match_when_negated` checked against an unmatching value

Can you check to make sure all those cases are covered, and add specs as needed?